docs/proposals/11-apiserver-network-proxy.md

```diff
@@ -0,0 +1,108 @@
+
+We want to replace this:
+
+`APIServer | VPN-seed ---> internet ---> LB --> VPN-Shoot (4314) --> Pods | Nodes | Services`
+
+With this:
+
+`APIServer <-> Proxy-Server <--- internet <--- Proxy-Agent --> Pods | Nodes | Services`
+
+
+### API Server Network Proxy
+
+To solve this issue we can utilize the [apiserver-network-proxy](apiserver-network-proxy) upstream implementation. Which provides a reference implementation for a reverse streaming server. The way it works is as follows:
+
+- Proxy agent connects to proxy server to establish a sticky connection.
+- Traffic to the proxy server gets then re-directed to the agent which forwards the traffic to in-cluster components.
+
+The initial motivation for the apiserver-network-proxy project is to get rid of provider-specific implementations that reside in the API-server (e.g., SSH), but it turns out that
+it has other interesting use-cases such as data-plane connection decoupling, which is the main use-case for this proposal.
+
+Starting **Kubernetes 1.18** its possible to make use of an ```--egress-selector-config-file``` flag, this helps point the API-server to traffic hook points based on traffic direction, for example, in the config below, the API server would have to forward all cluster related traffic (e.g., logs, port-forward, exec, ...etc.) to the **proxy-server** which then knows how to forward traffic to the shoot. For the rest of the traffic e.g., API server to ETCD or other control-plane components `direct` is used which means legacy routing method, i.e., by-pass the proxy.
+
+```yaml
+  egress-selector-configuration.yaml: |-
+    apiVersion: apiserver.k8s.io/v1alpha1
+    kind: EgressSelectorConfiguration
+    egressSelections:
+    - name: cluster
+      connection:
+        proxyProtocol: httpConnect
+        transport:
+          tcp:
+            url: https://proxy-server:8131
+    - name: master
+      connection:
+        proxyProtocol: direct
+    - name: etcd
+      connection:
+        proxyProtocol: direct
+```
+
+### Challenges
+
+One challenge remains to completely eliminate the need for a VPN connection. In today's Gardener setup, each control-plane has a Prometheus instance that directly scrapes cluster components such as CoreDNS, Kubelets, cadvisor, etc. This works because in addition to the VPN side car attached to the API server pod, we have another one attached to prometheus which knows how to forward traffic to these endpoints. Once the VPN is eliminated, it is required to find other means to forward traffic to these components.
+
+
+### Solutions
+
+There are currently two ways to solve this problem:
+
+- Attach a port-forwarder side-car to prometheus.
+- Utilize the proxy subresource on the API server.
+
```

Comment:
What about simply installing ip routes in the Prometheus network namespace (using a sidecar) that redirects traffic to the shoot networks to the proxy in the seed? Basically, this would be the same solution like for the kube-apiserver with the only difference that it doesn't/can't reuse the upstream network egress configuration.

Reply:
thought about it, but it wouldn't work that easily, for that you would need an additional sidecar that acts as proxy to the proxy, so basically we have another level of re-direction and another component that we have maintain.